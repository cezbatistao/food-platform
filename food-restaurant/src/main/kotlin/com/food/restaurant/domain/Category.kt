package com.food.restaurant.domain

data class Category(
        val id: Long?,
        val code: String,
        val description: String
) {
    constructor(code: String, description: String) :
            this(null, code, description)
}
