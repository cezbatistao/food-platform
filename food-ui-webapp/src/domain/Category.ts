export default class Category {
  _id: string;
  _description: string;

  constructor(id: string, description: string) {
    this._id = id;
    this._description = description;
  }

  get id(): string {
    return this._id;
  }

  get description(): string {
    return this._description;
  }
}