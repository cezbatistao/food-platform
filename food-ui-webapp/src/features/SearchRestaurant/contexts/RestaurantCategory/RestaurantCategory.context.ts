import { createContext } from "react";

export type CategoryContextState = {
  category: string;
  setCategory: (name: string) => void;
};

const contextDefaultValue: CategoryContextState = {
  category: '', 
  setCategory: () => {}
};
  
export const CategoryContext = createContext<CategoryContextState>(
  contextDefaultValue
);

export default CategoryContext;