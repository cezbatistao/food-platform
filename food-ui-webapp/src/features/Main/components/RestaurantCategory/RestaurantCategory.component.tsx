import { useEffect, useState, useContext } from 'react';
import { useSelector, useDispatch } from 'react-redux';
import { useLocation, useHistory } from "react-router-dom";
import queryString from 'query-string';
import { Location } from 'history';

import { fetchCategories } from '../../../../services/Category.services';
import CategoryContext from '../../contexts/RestaurantCategory/RestaurantCategory.context';
import Category from '../../../../services/Category.model';

const RestaurantCategory = () => {

  const [categories, setCategories] = useState<Category[]>([]);
  const [loading, setLoading] = useState<boolean>(true);

  const { setCategory } = useContext(CategoryContext);

  useEffect(() => {
    const getCategories = async () => {
      const allCategories: Category[] = await fetchCategories();
      setCategories(allCategories);
      setLoading(false);
    }

    setLoading(true);
    getCategories();
  }, []);

  const handleChangeCategory = (event: React.ChangeEvent<HTMLSelectElement>) => {
    const categorySelectedFromSelect = event.target.value as string;
    setCategory(categorySelectedFromSelect)
    // setCategorySelected(categorySelectedFromSelect);
    // dispatch(getRestaurantsByCategory(categorySelectedFromSelect));
    // var refresh = `${window.location.protocol}//${window.location.host + window.location.pathname}?category=${categorySelectedFromSelect}`;
    // window.history.pushState({ path: refresh }, '', refresh);
  }

  return (
    <select 
      data-test="select_category" 
      className="form-select" 
      aria-label="Selecione uma categoria" 
      onChange={ handleChangeCategory }
      defaultValue=''
      // value= { categorySelected }
    >
      <option value='' disabled selected>Selecione uma categoria</option>
      {!loading && 
        categories.map(category => {
          return (
            <option 
              key={ category.uuid } 
              value={ category.code }
            >
              { category.description }
            </option>
          );
        })
      }
    </select>
  );
}

export default RestaurantCategory;