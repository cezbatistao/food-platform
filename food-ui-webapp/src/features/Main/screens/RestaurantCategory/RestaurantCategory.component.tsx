import { useEffect, useState, useContext } from 'react';
import { useLocation } from "react-router-dom";
import queryString from 'query-string';
import { Location } from 'history';

import { fetchCategories } from '../../../../services/Category.services';
import CategoryContext from '../../contexts/RestaurantCategory/RestaurantCategory.context';
import Category from '../../../../services/Category.model';

const RestaurantCategory = () => {

  const [categories, setCategories] = useState<Category[]>([]);
  const [loading, setLoading] = useState<boolean>(true);

  const { category, setCategory } = useContext(CategoryContext);

  const location = useLocation();

  useEffect(() => {
    fetchCategories()
      .then(categories => {
        setCategories(categories);
        setLoading(false);
      });

    const categoryQueryString = getCategoryFromQueryString(location);
    setCategory(categoryQueryString);
  }, [location]);

  const getCategoryFromQueryString = (location: Location<unknown>) => {
    const parsed = queryString.parse(location.search);
    return parsed.category ? parsed.category as string : "";
  }

  const handleChangeCategory = (event: React.ChangeEvent<HTMLSelectElement>) => {
    const categorySelected = event.target.value as string;
    setCategory(categorySelected);
    var refresh = `${window.location.protocol}//${window.location.host + window.location.pathname}?category=${categorySelected}`;
    window.history.pushState({ path: refresh }, '', refresh);
  }

  return (
    <select 
      data-testid="select-category" 
      className="form-select" 
      aria-label="Selecione uma categoria" 
      onChange={ handleChangeCategory }
      defaultValue={ category }
    >
      <option data-testid="no-category" value="" disabled>Selecione uma categoria</option>
      {!loading && 
        categories.map(category => {
          return (
            <option 
              data-testid={ category.code } 
              key={ category.uuid } 
              value={ category.code }
            >
              { category.description }
            </option>
          );
        })
      }
    </select>
  );
}

export default RestaurantCategory;