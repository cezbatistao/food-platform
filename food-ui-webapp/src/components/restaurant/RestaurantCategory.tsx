import { makeStyles } from '@material-ui/core/styles';

import FormControl from '@material-ui/core/FormControl';
import Select from '@material-ui/core/Select';
import MenuItem from '@material-ui/core/MenuItem';
import InputLabel from '@material-ui/core/InputLabel';

import { useEffect } from 'react';
import { useSelector, useDispatch } from 'react-redux';

import Category from '../../domain/Category';
import { getCategories } from '../../gateway/actions/category.actions';
import { getRestaurantsByCategory } from '../../gateway/actions/restaurant.actions';
import { RootState } from '../../gateway';

const useStyles = makeStyles((theme) => ({
  formControl: {
    margin: theme.spacing(1),
    minWidth: 120,
  },
  paper: {
    padding: theme.spacing(2),
    display: 'flex',
    overflow: 'auto',
    flexDirection: 'column',
  },
  fixedHeight: {
    height: 240,
  }
}));

type CategoryState = {
  categories: Category[], 
  loading: boolean
}

const RestaurantCategory = () => {

  const handleChangeCategory = (event: React.ChangeEvent<{ value: unknown }>) => {
    const categorySelected = event.target.value as string;
    dispatch(getRestaurantsByCategory(categorySelected));
  }

  const dispatch = useDispatch();
  useEffect(() => {
    dispatch(getCategories());
  }, [dispatch]);

  const { loading, categories }: CategoryState = useSelector((
    state: RootState
  ) => state.category);

  const classes = useStyles();

  return (
    <div>
      <FormControl className={classes.formControl}>
        <InputLabel id="demo-simple-select-label">Categoria</InputLabel>
        <Select 
          labelId="demo-simple-select-label"
          id="demo-simple-select"
          onChange={handleChangeCategory}
        >
          {!loading && 
            categories.map(category => {
              return <MenuItem key={ category.id } value={ category.id }>{ category.description }</MenuItem>;
            })
          }
        </Select>
      </FormControl>
    </div>
  );
}

export default RestaurantCategory;