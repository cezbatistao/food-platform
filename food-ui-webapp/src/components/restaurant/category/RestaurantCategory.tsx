import { useDispatch } from 'react-redux';

import Category from '../../../domain/Category';
import { getRestaurantsByCategory } from '../../../gateway/actions/restaurant.actions';

type Props = {
  loading: boolean,
  categories: Category[]
}

const RestaurantCategory = ({ loading, categories }: Props) => {
  
  const dispatch = useDispatch();

  const handleChangeCategory = (event: React.ChangeEvent<HTMLSelectElement>) => {
    // event.preventDefault();
    const categorySelected = event.target.value as string;
    dispatch(getRestaurantsByCategory(categorySelected));
  }

  return (
    <select className="form-select" aria-label="Default select example" onChange={ handleChangeCategory }>
      <option disabled selected>Selecione uma categoria</option>
      {!loading && 
        categories.map(category => {
          return (
            <option 
              key={ category.uuid } 
              value={ category.code }
            >
              { category.description }
            </option>
          );
        })
      }
    </select>
  );
}

export default RestaurantCategory;