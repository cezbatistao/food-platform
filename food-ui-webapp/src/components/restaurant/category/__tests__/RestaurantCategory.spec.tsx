import { shallow } from 'enzyme';

import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';

import RestaurantCategory from "../RestaurantCategory";
import Category from "../../../../domain/Category";

import categoriesResponseJson from "../../../../../dependencies/stubby/mocks/response/categories_response.json";

const mockUseDispatch = jest.fn();
jest.mock('react-redux', () => ({
  useSelector: jest.fn(),
  useDispatch: () => mockUseDispatch
}));

const mockGetRestaurantsByCategory = jest.fn();
jest.mock('../../../../gateway/actions/restaurant.actions', () => ({
  getRestaurantsByCategory: (category: string) => mockGetRestaurantsByCategory(category)
}));

describe('RestaurantCategory', () => {

  beforeEach(() => {
    mockUseDispatch.mockClear();
    mockGetRestaurantsByCategory.mockClear();
  });

  it("renders the categories from restaurant", () => {
    mockUseDispatch.mockReturnValue(jest.fn());
    mockGetRestaurantsByCategory.mockReturnValue(jest.fn());

    const wrapper = shallow(
      <RestaurantCategory 
        loading={ false } 
        categories={ categoriesResponseJson.data as Category[] }
      />
    );

    expect(wrapper.find('select').length).toBe(1);
    expect(wrapper.find('select option').length).toBe(6);
    
    expect(mockUseDispatch).toHaveBeenCalledTimes(0);
    expect(mockGetRestaurantsByCategory).toHaveBeenCalledTimes(0);

    expect(wrapper).toMatchSnapshot();
  });

  it("renders empty categories", () => {
    mockUseDispatch.mockReturnValue(jest.fn());
    mockGetRestaurantsByCategory.mockReturnValue(jest.fn());

    const wrapper = shallow(
      <RestaurantCategory 
        loading={ false } 
        categories={ [] }
      />
    );

    expect(wrapper.find('select').length).toBe(1);
    expect(wrapper.find('select option').length).toBe(1);

    expect(mockUseDispatch).toHaveBeenCalledTimes(0);
    expect(mockGetRestaurantsByCategory).toHaveBeenCalledTimes(0);

    expect(wrapper).toMatchSnapshot();
  });

  it("should dispatch a getRestaurantsByCategory when select a category of restaurant", () => {
    const valueToSelect = 'pizza';

    mockUseDispatch.mockReturnValue(jest.fn());
    mockGetRestaurantsByCategory.mockReturnValue(jest.fn());

    const wrapper = shallow(
      <RestaurantCategory 
        loading={ false } 
        categories={ categoriesResponseJson.data as Category[] }
      />
    );
    
    wrapper.find('select').simulate('change', {
      target: { value: valueToSelect }
    });

    expect(mockUseDispatch).toHaveBeenCalledTimes(1);
    expect(mockGetRestaurantsByCategory).toHaveBeenCalledTimes(1);

    expect(mockGetRestaurantsByCategory.mock.calls[0][0]).toBe(valueToSelect)
  });
});