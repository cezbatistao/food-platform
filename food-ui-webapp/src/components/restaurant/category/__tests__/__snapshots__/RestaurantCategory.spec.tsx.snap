// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RestaurantCategory renders empty categories 1`] = `
<select
  aria-label="Default select example"
  className="form-select"
  onChange={[Function]}
>
  <option
    disabled={true}
    selected={true}
  >
    Selecione uma categoria
  </option>
</select>
`;

exports[`RestaurantCategory renders the categories from restaurant 1`] = `
<select
  aria-label="Default select example"
  className="form-select"
  onChange={[Function]}
>
  <option
    disabled={true}
    selected={true}
  >
    Selecione uma categoria
  </option>
  <option
    key="0eda194c-827c-4254-ada8-214115310fc6"
    value="pizza"
  >
    Pizzaria
  </option>
  <option
    key="cc0cd1b9-4217-498d-8da8-661de03b86e5"
    value="hamburguer"
  >
    Hamb√∫rguer
  </option>
  <option
    key="577a9962-e9b7-46a9-abc8-211e42f6f1aa"
    value="japonesa"
  >
    Japonesa
  </option>
  <option
    key="67e92973-f1ff-4cc3-b2c0-6485939ae442"
    value="vegetariana"
  >
    Vegetariana
  </option>
  <option
    key="2358fb2d-f0dc-4a00-8b55-6fbbce158e91"
    value="brasileira"
  >
    Brasileira
  </option>
</select>
`;
